Add MVP for cli

THenry14 opened the pull request:
### Checklist
- [x] Relevant issue is linked
- [ ] Docs updated/issue for docs created
- [ ] Added relevant tests

Closes https://github.com/software-mansion/protostar/issues/1948
Closes https://github.com/software-mansion/protostar/issues/2104
Closes https://github.com/software-mansion/protostar/issues/2118

THenry14 committed:
[skip ci]Add draft for cli

MaksymilianDemitraszek commented on protostar-cli/src/lib.rs:
```diff
@@ -0,0 +1,82 @@
+use url::Url;
+
+#[derive(Deserialize, Serialize)]
+struct ParsedAccount {
+    private_key: String,
+    public_key: String,
+    address: String,
+}
```
```suggestion
struct Account {
    private_key: String,
    public_key: String,
    address: String,
}
```
probably a better name because it being represented as a structure implies it being parsed

MaksymilianDemitraszek commented on protostar-cli/src/main.rs:
```diff
@@ -0,0 +1,99 @@
+    let cli = Cli::parse();
+
+    let network_name = cli.network.unwrap_or_else(|| {
+        env::var("STARKNET_NETWORK").unwrap_or_else(|_| {
```
We probably shouldn't support env variable it could be specified in the scarb config instead.

MaksymilianDemitraszek commented on protostar-cli/src/main.rs:
```diff
@@ -0,0 +1,99 @@
+    network: Option<String>,
+
+    #[command(subcommand)]
+    command: Option<Commands>,
```
What is the reason behind it being optional?

THenry14 committed:
Add declare command + rebase

war-in commented on protostar-cli/src/starknet_commands/declare.rs:
```diff
@@ -0,0 +1,49 @@
+    sierra_contract_path: &Utf8PathBuf,
+    casm_contract_path: &Utf8PathBuf,
+    account: &mut SingleOwnerAccount<&JsonRpcClient<HttpTransport>, LocalWallet>,
+) -> Result<DeclareTransactionResult, Error> {
```
As @cptartur said, Error is unnecessary here

cptartur commented on protostar-cli/src/lib.rs:
```diff
@@ -0,0 +1,80 @@
+    Ok(account)
+}
+
+pub fn get_account<'a>(
+    name: &str,
+    accounts_file_path: &Utf8PathBuf,
+    provider: &'a JsonRpcClient<HttpTransport>,
+    network: &Network,
+) -> Result<SingleOwnerAccount<&'a JsonRpcClient<HttpTransport>, LocalWallet>> {
+    // todo: #2113 verify network with provider
+    let account_info = get_account_info(name, network.get_value(), accounts_file_path)?;
+    let signer = LocalWallet::from(SigningKey::from_secret_scalar(
+        FieldElement::from_hex_be(&account_info.private_key).unwrap(),
+    ));
+    let address = FieldElement::from_hex_be(&account_info.address).unwrap();
+    let mut account = SingleOwnerAccount::new(provider, signer, address, network.get_chain_id());
+
+    Ok(account)
+}
```
I'd replace these `unwrap` with `?` or some other handling. You can consider adding `context` to these errors though it's not necessary.

cptartur commented on protostar-cli/src/lib.rs:
```diff
@@ -0,0 +1,80 @@
+}
+
+fn get_account_info(name: &str, chain_id: &str, path: &Utf8PathBuf) -> Result<Account> {
+    let accounts: Value = serde_json::from_str(&fs::read_to_string(path)?)?;
```
Why are we doing deserialization to `Value` first instead of defining a struct for that and using it directly?

THenry14 replied:
Right, changed it to use Account struct

war-in committed:
[cli] add call command (#2110)

### Checklist
- [x] Relevant issue is linked
- [ ] Docs updated/issue for docs created
- [ ] Added relevant tests

Closes #2054

cptartur commented on protostar-cli/src/starknet_commands/declare.rs:
```diff
@@ -0,0 +1,49 @@
+    account: &mut SingleOwnerAccount<&JsonRpcClient<HttpTransport>, LocalWallet>,
+) -> Result<DeclareTransactionResult, Error> {
+    let contract_definition: SierraClass = {
+        let file_contents = std::fs::read(sierra_contract_path)
+            .map_err(|err| anyhow::anyhow!("Failed to read contract file: {}", err))?;
```
```suggestion
        let file_contents = std::fs::read(sierra_contract_path)
            .with_context(|| format!("Failed to read contract file: {}", err))?;
```

cptartur replied:
Nobody asked, but about `context` vs `with_context`. `with_context` is lazily evaluated so it runs only on error, so you won't be doing costly string formatting if not necessary.

cptartur reviewed (commented):
Couple of nits, but the logic overall looks good.

cptartur committed:
Add invoke to CLI (#2106)

Closes #2053 

### Checklist
- [ ] Relevant issue is linked
- [ ] Docs updated/issue for docs created
- [ ] Added relevant tests

THenry14 committed:
Rename project to cast

THenry14 committed:
review changes

THenry14 committed:
Add clippy as code checker

THenry14 committed:
Apply clippy suggestions

THenry14 committed:
Use Account struct for accounts file deserialization

cptartur commented on cast/.cargo/config.toml:
```diff
@@ -0,0 +1,2 @@
+[alias]
+lint = "clippy --all-targets --all-features -- --no-deps -W clippy::pedantic -A clippy::missing_errors_doc -A clippy::missing_panics_doc -A clippy::default_trait_access"
```
Newline missing
